### PR TITLE
BUG: Use MST algorithm instead of SLINK for single linkage clustering

### PR DESCRIPTION
**Pull request by nmayorov**
Following #5960 I replaced SLINK algorithm with MST. It was rather straightforward to implement, except that post-labeling algorithm had to be changed slightly as well.

Timings show that MST is slightly faster on a random input and NN-CHAIN is about 2 times slower (so using MST makes sense).

Tests with ties (the very reason to replace SLINK) were taken from #5960.

### CODE REVIEW

**Comment by rgommers** [comment]
Interesting that the Mullner paper has 54 citations already but was never published in a journal.

This PR implements the method originally by  Rohlf (1973), so citing that paper would be good as well.

**Comment by rgommers** [comment]
@nmayorov it would be good to cherry-pick https://github.com/scipy/scipy/pull/5960/commits/b82937e6e63b so you preserve the authorship info.

**Comment by nmayorov** [comment]
@rgommers I think it's done.

File: scipy/cluster/_hierarchy.pyx [hunk 1]
```diff
@@ -914,16 +939,17 @@ cdef class LinkageUnionFind:
...
     cdef int i, x, y, x_root, y_root
     for i in range(n - 1):
         x, y = int(Z[i, 0]), int(Z[i, 1])
         x_root, y_root = uf.find(x), uf.find(y)
-        uf.merge(x, y)
         if x_root < y_root:
             Z[i, 0], Z[i, 1] = x_root, y_root
         else:
             Z[i, 0], Z[i, 1] = y_root, x_root
+        Z[i, 3] = uf.merge(x_root, y_root)
```
Comment (rgommers) [inline scipy/cluster/_hierarchy.pyx]: This is a fix in ``label`` that affects ``nn_chain`` as well right? And the issue is covered by one of the new tests?

**Review by rgommers — APPROVED** [review]
LGTM

File: scipy/cluster/_hierarchy.pyx (hunk 1 above)
Comment (nmayorov) [inline scipy/cluster/_hierarchy.pyx]: This change doesn't affect `nn_chain`, but is required for `mst_linkage` to label clusters correctly. It is not related to the ties issue.

So the new variant is more general and the original version relies on the order of merges in `nn_chain`, which is quite nontrivial thing. Partially my confidence is based on the extensive test suite available in scipy. I also run couple of comparisons with fastcluster on random inputs.

Comment (rgommers) [inline scipy/cluster/_hierarchy.pyx]: > the original version relies on the order of merges in ``nn_chain``

ah thanks, that wasn't clear to me.

**Comment by rgommers** [comment]
In it goes then. Thanks @nmayorov!

**Comment by rgommers** [comment]
won't do an 0.18.x release anymore, so removing the backport label

**Pull request merged**